### DISCUSSION TITLE
Improve the speed on a mobile device of a YOLOv8 OD model

### DISCUSSION THREAD

**Discussion by zfrankowski** [ultralytics/ultralytics#21662] [General]
Good afternoon everyone!

   I have a YOLOv8 model that I have trained to detect street signs and am using it on a Samsung Galaxy S24 Ultra.  It only detects if I'm driving really slow and even then it takes about 1.5s to detect a sign.  Is there anyone way of speeding this up?  I don't really want to drop the image size as it's an object detection model and that might lose detection accuracy.  I exported it to tflite and am graciously using (instead of reinventing the wheel) https://github.com/surendramaran/YOLO Android Studio project.

   I'm newish to YOLO and would appreciate any comments or feedback!!

Thanks - Zack

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zfrankowski, thanks for starting the discussion and for using Ultralytics 🚀 This is an automated response to help you get quick guidance while an Ultralytics engineer reviews your post and assists you soon.

We recommend a visit to the Docs for general usage tips and examples:
- Python: https://docs.ultralytics.com/usage/python/
- CLI: https://docs.ultralytics.com/usage/cli/
- Main Docs: https://docs.ultralytics.com/

If this is a 🐛 Bug Report, please provide a minimum reproducible example (MRE) to help us debug it: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please share as much detail as possible, including:
- Dataset samples (a few images)
- Training command and logs
- Export details (format, args)
- Inference environment and logs
And verify you’re following our Tips for Best Training Results: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord for real-time chat 🎧: https://discord.com/invite/ultralytics
- Discourse for in-depth threads: https://community.ultralytics.com/
- Subreddit for community tips: https://www.reddit.com/r/Ultralytics/

## Upgrade

Please upgrade to the latest ultralytics package with all requirements in a Python>=3.8 environment with PyTorch>=1.8 to ensure you’re on the newest version:
```bash
pip install -U ultralytics
```
Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Hi Zack — to speed up on S24 Ultra, export your model to TFLite with FP16 or INT8 and enable NMS at export, then benchmark different delegates (GPU, NNAPI, Hexagon) as performance varies by device; if possible, try a smaller variant or move to YOLO11n for better speed/accuracy, and use the Ultralytics HUB App to quickly sanity‑check on‑device performance with accelerated TFLite. See the TFLite export guide for FP16/INT8 and NMS, the Android App docs on quantization and delegates, and the YOLO11 model family.

**Reply by zfrankowski** [reply]
Hi Glenn - Thanks so much for the information.  I will definitely try that!

Zack

**Reply by glenn-jocher** [reply]
Great—when you test, try NNAPI and GPU delegates on the S24 Ultra, set the TFLite Interpreter to 4–8 threads, export with nms=True and the same imgsz you’ll use on-device, and for INT8 use a small calibration set via data and fraction; also ensure the app reads the model’s built-in NMS outputs to avoid extra CPU post-processing—please share which combo gives you the best FPS.

**Reply by zfrankowski** [reply]
I'm not sure how to program delegates so I'll have to google but I will definitely let everyone know!

**Reply by glenn-jocher** [reply]
No worries—on Android you select a TFLite delegate via Interpreter.Options (GpuDelegate or NnApiDelegate) and set threads; the TensorFlow Lite GPU delegate quickstart and the Android NNAPI overview show how, and you can quickly test performance without coding using the Ultralytics Android App guide.

**Reply by zfrankowski** [reply]
Great, thanks Glenn!  I will definitely check it out!

**Reply by glenn-jocher** [reply]
Perfect—when you test, A/B GPU vs NNAPI (Hexagon) and FP16 vs INT8 using the Ultralytics Android App guide and export via the TFLite export guide; share results and we’ll help tune.

**Reply by zfrankowski** [reply]
Sounds great, I really appreciate the help!  Hopefully I can get it going - my Android Studio skills are rusty but I'll try my best!!

**Reply by zfrankowski** [reply]
I got the nnapi delegate added however it says the GPU delegate on my cell phone isn't available.  I even asked Gemini for code verification it didn't seem to find anything wrong.  I'm going to go try it in the real world now.  I'm not sure how to actually benchmark test it though - I'm sorry I sound stupid, still new to this. :(

** UPDATE: I noticed that the inference time went down from 1.5s running on CPU to 0.5s running on the NNANPI delegate.  it's a start :)  I find that if I drive at 30km/h it will detect signage but any faster it doesn't.

**Comment by Y-T-G** [comment]
What's the size of the model?

Ultralytics Flutter App uses GPUDelegate, and you can easily get under 30ms inference with that for nano models.

https://github.com/ultralytics/yolo-flutter-app

You can test the app from Play Store by downloading Ultralytics HUB.

**Reply by zfrankowski** [reply]
Hi Y-T-G - The current model is 22Mb.  I'm training a slightly larger model, but will definitely try it in Flutter and in the Ultralytics Hub.  I will post my results soon.  I'm going to train tomorrow on the new model.

Thanks - Zack

**Reply by glenn-jocher** [reply]
22 MB is mid-size; for real-time on mobile go smaller, not larger—train a YOLO11n/s, export to TFLite with FP16 or INT8 and NMS, then A/B NNAPI vs GPU using the Ultralytics Android App. To gain speed without losing much accuracy, limit classes/max_det to your signs and crop the camera ROI instead of increasing imgsz. See the TFLite export guide and the Ultralytics Android App docs for the exact steps.

**Reply by zfrankowski** [reply]
Ok will definitely try that!!  Thanks again Glenn and Y-T-G!  Will post results here soon.  I'm excited to see that there is a Flutter option as I'm very familiar with that. :)

**Comment by zfrankowski** [comment]
Hi all!  I was able to achieve 20-30fps using the Flutter option with my model, upon initial load it downloads the yolov11 model and it's detecting common objects plus what I have in my model.  Is there any way to turn the common object detection off in the Flutter app and only detect what's in the trained model?  Here is my main.dart file.

```
void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(title: 'YOLO Example', home: YOLOScreen());
  }
}

class YOLOScreen extends StatelessWidget {
  const YOLOScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('YOLO Detection')),
      body: YOLOView(
        modelPath: 'best_float16-2025-08-19-y11s',
        task: YOLOTask.detect,
        onResult: (results) {
          for (final result in results) {
            debugPrint('${result.className}: ${result.confidence}');
          }
        },
      ),
    );
  }
}

// Single image inference example
class SingleImageExample extends StatefulWidget {
  const SingleImageExample({super.key});

  @override
  State<SingleImageExample> createState() => _SingleImageExampleState();
}

class _SingleImageExampleState extends State<SingleImageExample> {
  YOLO? _yolo;

  @override
  void initState() {
    super.initState();
    _initializeYOLO();
  }

  Future<void> _initializeYOLO() async {
    _yolo = YOLO(
      modelPath: 'best_float16-2025-08-19-y11s',
      task: YOLOTask.detect,
    );
    await _yolo!.loadModel();
  }

  @override
  Widget build(BuildContext context) {
    return Container(); // Simplified for brevity
  }
}

```

**Reply by glenn-jocher** [reply]
Good progress—what you’re seeing likely means the app is loading the default COCO model/labels instead of your custom one. Point YOLOView to your actual .tflite asset (use the full filename with .tflite, e.g., assets/models/best_float16-2025-08-19-y11s.tflite), bundle the matching labels file for your classes, and ensure any auto-download/default model paths are disabled. You can also filter client‑side to show only your classes:

```dart
body: YOLOView(
  modelPath: 'assets/models/best_float16-2025-08-19-y11s.tflite',
  task: YOLOTask.detect,
  onResult: (results) {
    const allowed = {'stop_sign', 'speed_limit', 'yield'}; // your classes
    for (final r in results.where((r) => allowed.contains(r.className))) {
      debugPrint('${r.className}: ${r.confidence}');
    }
  },
),
```

If needed, re‑export your model to TFLite so you have the correct file to bundle; see the TFLite integration guide for steps and options: See the Ultralytics TFLite guide for details.

**Reply by zfrankowski** [reply]
Hmmmm, I'm stuck.  I can't seem to get it to disable the auto download of the yolo11n file without it not loading my model.  I specified my model in the modelPath and it works, but just in conjunction with yolo11n.  I think I'm going to leave it for today and retry tomorrow with fresh eyes.  In the mean time any help is always appreciated, I can't say enough how much you guys rock!  I'm sorry to be a pain!!

**Reply by glenn-jocher** [reply]
Good progress. The auto-download comes from the YOLOView convenience widget; to force only your custom model, bundle it as an asset and load it directly:

1) Add to pubspec.yaml:
flutter:
  assets:
    - assets/models/best_float16-2025-08-19-y11s.tflite
    - assets/models/labels.txt

2) Uninstall/reinstall the app to clear any cached yolo11n.

3) Bypass YOLOView and load the model yourself:
final yolo = YOLO(modelPath: 'assets/models/best_float16-2025-08-19-y11s.tflite', task: YOLOTask.detect);
await yolo.loadModel();

If it still pulls a default model, please share the package name/version, your pubspec assets section, and the app startup log so we can reproduce and advise.

**Comment by zfrankowski** [comment]
Hi Glenn - Sorry that didn't work.  I've attached a copy of my main.dart

Attached are the debug logs...
[logs.txt](https://github.com/user-attachments/files/21918877/logs.txt)
[main.txt](https://github.com/user-attachments/files/21918880/main.txt)
[pubspec.yaml](https://github.com/user-attachments/files/21918904/pubspec.yaml)

Thanks so much in advance!!
Zack

**Reply by zfrankowski** [reply]
Ok - I got a work around - I put my model on my webserver and changed the yolov11n download site to that and it seems to work now.  The model is slow now at 5-6fps down from 30fps with the yolov11n model included so I need to work on that.

**Reply by glenn-jocher** [reply]
Nice progress—5–6 FPS usually means your custom .tflite is either larger than the bundled nano, exported as FP32/INT8 (GPU can’t run INT8), or using a bigger imgsz; re‑export your custom model as YOLO11n FP16 with fused NMS at 640 and confirm the log shows “TfLiteGpuDelegateV2” is created:

from ultralytics import YOLO
YOLO('path/to/best.pt').export(format='tflite', half=True, nms=True, imgsz=640)

If it still falls back off‑GPU, try NNAPI; the quantization/delegate trade‑offs are summarized in the Android guide: See the Ultralytics Android App docs.

**Reply by zfrankowski** [reply]
Hi Glenn - Sounds good I will give that a shot.

Thanks! Zack

**Reply by glenn-jocher** [reply]
Sounds good—after re-export, please confirm the on-device log shows “Created TfLiteGpuDelegateV2” (FP16 + fused NMS) and, if FPS is still low, try cropping the camera ROI and limiting classes/max_det to your signs; step-by-step export and Android delegate guidance are in the [TFLite export guide](https://docs.ultralytics.com/integrations/tflite/) and the [Android App docs](https://docs.ultralytics.com/hub/app/android/).

**Reply by zfrankowski** [reply]
Thanks Glenn I will try that on Monday unfortunately I'm busy today.  I appreciate your help so much!!

**Reply by glenn-jocher** [reply]
No rush—when you resume, profile a release build (not debug), set the camera capture/preview resolution close to your model `imgsz` (e.g., 640) and add a brief warm‑up; if it’s still slow, re‑train/export a YOLO11n FP16 with fused NMS for mobile; see the [mobile and edge deployment FAQ](https://docs.ultralytics.com/help/FAQ/) for extra tips.

**Comment by zfrankowski** [comment]
Hi Glenn - I'm still working on the model.  It'll be a couple weeks til I get back at it if you can keep this thread open. :)

Thanks - Zack

**Reply by glenn-jocher** [reply]
Thanks Zack—no problem, we’ll keep this thread open; if it auto‑closes later, please start a new discussion and link back here. When you’re back, share the release-build log line showing which delegate initialized plus your `.tflite` size/export args; for a quick desktop baseline you can run:

```python
from ultralytics.utils.benchmarks import benchmark
benchmark(model="path/to/best.pt", imgsz=640, device=0)
```

See the [Benchmark mode docs](https://docs.ultralytics.com/modes/benchmark/) for details.
